Spatially Resolved Thermochemical analysis of Molten Salt Fast Reactor (#316)
idaholab/virtual_test_bed: doc/content/msr/msfr/thermo_model.md

> # MSFR Griffin-Pronghorn-Thermochimica Model

Reviewer: ```suggestion
# MSFR Spatially-resolved thermo-chemistry model using Thermochimica
```
I would change this because you are not running Griffin nor pgh

Author: Accepted into push.

Next the `output_element_potentials` produces the chemical potentials of all elements in the solution so the fluorine or iodine potential can be visualized throughout the reactor given the temperature, pressure, and element field.

Lastly, if a stable ideal_gas phase is not present, the `output_vapor_pressures` can also be selected which will determine the vapor pressures of elements in the liquid MSFL phase given there is a liquid-gas interface present.

It should be noted that this block creates AuxVariables of these inputs and ouputs defined here and can be manipulated and exported to an exodus file for viewing.

#### Chemical Composition

Next, the `ChemicalComposition` block is used to set up the Thermochimica calculation and is described in more detail [here](https://mooseframework.inl.gov/source/actions/ChemicalCompositionAction.html). First the modified MSTDB V2.0 with noble gas data file is specified, and the temperature, pressure, and mass unites as chosen.

!listing msr/msfr/thermochemistry/thermo.i block=ChemicalComposition

> Next, the auxiliary variables of the temperature and pressure are specified as `tfuel_nod` and `pressure_nod` respetively. It should be noted that the element Auxiliary variables were already declared in the `GlobalParams` block. Lastly, a `reinitilization_type` is specified to use cached Thermochimica solutions to speed up calculations.

Reviewer: ```suggestion
Next, the auxiliary variables of the temperature and pressure are specified as `tfuel_nod` and `pressure_nod` respectively. It should be noted that the element Auxiliary variables were already declared in the `GlobalParams` block. Lastly, a `reinitialization_type` is specified to use cached Thermochimica solutions to speed up calculations.
```

Author: Accepted into new push.

!listing msr/msfr/thermochemistry/thermo.i block=AuxVariables

Currently, the implementation of Thermochimica in MOOSE computes nodal values, so these finite volume variables must be converted to nodal variables which are the `tfuel_nod` and `pressure_nod` variables respectively. These variables are then used in the `ChemicalComposition` block.

#### Auxiliary Kernels

Correspondingly, the `AuxiliaryKernels` block sets up functors which will operate on the auxiliary variables. In this case, since a conversion from the finite volume variables to nodal variables is required, the `ProjectionAux` is used to perform this transformation. This is executed on `Initial` so that the nodal version of the variables are ready for the Thermochimica calculation.

!listing msr/msfr/thermochemistry/thermo.i block=AuxKernels

#### Initial Conditions

Next, the `Initial Conditions` block sets up the initial conditions for various Auxiliary variables. In this case, the element variables are set through a `FunctionIC` which calculates a homogenous molar distribution from a depletion calculation and will be discussed shortly.

!listing msr/msfr/thermochemistry/thermo.i block=ICs

#### Problem

> The `Problem` block is used here to tell MOOSE that a finite element solve is not required, since a MOOSE-wrapped application via `UserObject` is being used instead. Additionally, the `allow_initial_conditions_with_restart` argument is used to allow for reading in the temperature and pressure variables from the multiphysics restart file.

Reviewer: ```suggestion
The `Problem` block is used here to tell MOOSE that a finite element solve is not required, since a MOOSE-wrapped application via `UserObject` is being used instead. Additionally, the `allow_initial_conditions_with_restart` argument is used to allow for reading in the temperature and pressure variables from the multiphysics restart file, while initializing other variables from the input file.
```

Author: Accepted into new push.

!listing msr/msfr/thermochemistry/thermo.i block=Problem

#### Postprocessors

Next, the largest block of the input, the `Postprocessors` block is used to calculate the homogenous element distributions from a previous depletion calculation in a CSV file. Additionally, various total elements or species of interest can be calculated here.

Here each nuclide density in $[atoms/(b-cm)]$ is read from a `VectorPostProcessor` named `reader` (discussed in the `VectorPostprocessors` block) at a specific burnup index. In order to visualize the depletion-driven, spatially-resolved thermochemical effect, the user can replace the `index = 0` value with a later index point to model how the solution changes throughout depletion.

> Admittedly, this is a clunky method of passing information, and current efforts are streamlining nuclide, element, and species accounting for MSR analyses within the MOOSE framework. This will allow for a more user friendly interface between online Griffin-depletion coupling with Thermochimica + Pronghorn for depletion-driven, spatially-resolved thermochemistry modeling and simulation. Therefore, this model will likely be updated in the near term future with a more efficient streamlined version.

Reviewer: ```suggestion
Admittedly, this is a clunky method of passing information, and current efforts are streamlining nuclide, element, and species accounting for MSR analyses within the MOOSE framework. This will allow for a more user friendly interface between online Griffin-depletion coupling with Thermochimica + Pronghorn for depletion-driven, spatially-resolved thermochemistry modeling and simulation. Therefore, this model will likely be updated in the near term with a more efficient streamlined version.
```

Author: Accepted into new push.
